Feb blog post (#1004)
Azure/azure-sdk: _posts/2020-02-11-february-release.md

---
title: Azure SDK Releases (February 2020)
layout: post
date: 11 Feb 2020
sidebar: releases_sidebar
repository: azure/azure-sdk
---

Welcome to the February release of the Azure SDK.  We have updated the following libraries:

* Azure Storage (Blobs and File Shares).

> * Event Hubs

Reviewer: The following libraries also GA'ed: Key Vault Certs (Python), Key Vault Keys (Python), Key Vault Secrets (Python),  Storage Que (Python and Java), Cosmos (JS), Identity (.NET and Python)

These are ready to use in your production applications.  You can find details of all released libraries on [our releases page](https://azure.github.io/azure-sdk/releases/latest/).

New preview releases:

* [Azure Storage Data Lake Files](https://docs.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-introduction?toc=%2fazure%2fstorage%2fblobs%2ftoc.json)
* [Azure Text Analytics](https://docs.microsoft.com/en-us/azure/cognitive-services/text-analytics/index)

> 

Reviewer: The following libraries also shipped previews: Key Vault Keys Java, Cosmos Java, Identity Java

## Logging 

One of the main advantages of the new Azure SDK is that common developer concerns are treated the same, irrespective of which client library you are using.  In this issue, we will tackle logging and show you how to enable logging on each platform.  

## .NET libraries

All .NET client libraries emit events to ETW (Event Tracing for Windows) via the `EventSource` class.  This system has been a part of the .NET framework for a long time.  Event sources allow you to use structured logging in your application code with a minimal performance overhead.

Although you can use out-of-process tools (such as **TODO: INSERT TOOLS HERE WITH LINKS**), a [core tenet of our libraries](https://azure.github.io/azure-sdk/general_implementation.html#general-logging-console-logger) includes the ability to easily send logs to the console.  When developing your app, you should be able to view the logs in real time without much overhead.  You can accomplish this with a one-liner at the top of your application:

```csharp
using AzureEventSourceListener listener = AzureEventSourceListener.CreateConsoleLogger();
```

## Java libraries

> The Java client libraries all use [SLF4J](https://www.slf4j.org/) under the covers for logging.  There are many mechanisms for [configuring SLF4J](https://dzone.com/articles/how-configure-slf4j-different) to get just the right logging for you application.  However, we are looking at the simplest mechanism today.  There is a built-in logger for temporary debugging purposes.  As long as you have not already included the SLF4J library, just set the `AZURE_LOG_LEVEL` environment variable to "verbose".  For example, in bash:

Reviewer 2: @conniey could you review and update this section today to incorporate your PR from 2 days ago: https://github.com/Azure/azure-sdk-for-java/pull/8031

Reviewer 3: This PR only affected cosmos' tests. Their implementation of slf4j was log4j 1.x. It isn't shipped as part of the product.

> The Java client libraries all use [SLF4J](https://www.slf4j.org/) under the covers for logging.  There are many mechanisms for [configuring SLF4J](https://dzone.com/articles/how-configure-slf4j-different) to get just the right logging for you application.  However, we are looking at the simplest mechanism today.  There is a built-in logger for temporary debugging purposes.  As long as you have not already included the SLF4J library, just set the `AZURE_LOG_LEVEL` environment variable to "verbose".  For example, in bash:

Reviewer 3: ```suggestion
The Java client libraries all use [SLF4J](https://www.slf4j.org/) under the covers for logging.  There are many mechanisms for [configuring SLF4J](https://dzone.com/articles/how-configure-slf4j-different) to get just the right logging for your application.  By default, there are no SLF4J implementation in the client library. Today, we are looking at the most straightforward mechanism, [the simple logger](http://www.slf4j.org/api/org/slf4j/impl/SimpleLogger.html). 
As long as you have not already included another SLF4J implementation, download and add [slf4j-simple 1.7.30.jar](https://search.maven.org/artifact/org.slf4j/slf4j-simple/1.7.30/jar) to your classpath. Then, set the `AZURE_LOG_LEVEL` environment variable to "verbose".  For example, in bash:
```

## .NET libraries

All .NET client libraries emit events to ETW (Event Tracing for Windows) via the `EventSource` class.  This system has been a part of the .NET framework for a long time.  Event sources allow you to use structured logging in your application code with a minimal performance overhead.

Although you can use out-of-process tools (such as [PerfView](https://github.com/Microsoft/perfview) or [dotnet trace](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-trace)), a [core tenet of our libraries](https://azure.github.io/azure-sdk/general_implementation.html#general-logging-console-logger) includes the ability to easily send logs to the console.  When developing your app, you should be able to view the logs in real time without much overhead.  You can accomplish this with a one-liner at the top of your application:

```csharp
using AzureEventSourceListener listener = AzureEventSourceListener.CreateConsoleLogger();
```

## Java libraries

> The Java client libraries all use [SLF4J](https://www.slf4j.org/) under the covers for logging.  There are many mechanisms for [configuring SLF4J](https://dzone.com/articles/how-configure-slf4j-different) to get just the right logging for your application.  By default, there is no SLF4J implementation in the client library. Today, we will show the easiest implementation - [the simple logger](http://www.slf4j.org/api/org/slf4j/impl/SimpleLogger.html).  Download and add [slf4j-simple 1.7.30.jar](https://search.maven.org/artifact/org.slf4j/slf4j-simple/1.7.30/jar) to your classpath (ensuring you do not have another SLF4J library in your app).  Set the `AZURE_LOG_LEVEL` environment variable to "verbose".  For example, in bash:

Reviewer 4: ```suggestion
The Java client libraries all use [SLF4J](https://www.slf4j.org/) under the covers for logging.  There are many mechanisms for [configuring SLF4J](https://dzone.com/articles/how-configure-slf4j-different) to get just the right logging for your application.  By default, there is no SLF4J implementation in the client library. Today, we will show the easiest implementation - [the simple logger](http://www.slf4j.org/api/org/slf4j/impl/SimpleLogger.html).  Download and add [slf4j-simple 1.7.28.jar](https://search.maven.org/artifact/org.slf4j/slf4j-simple/1.7.28/jar) to your classpath (ensuring you do not have another SLF4J library in your app).  Set the `AZURE_LOG_LEVEL` environment variable to "verbose".  For example, in bash:
```

Will want to use `1.7.28` as that matches the version of `slf4j-api` we use as a dependency, will reduce chances a customer runs into version mismatch issues.

CC: @conniey

```bash
export AZURE_LOG_LEVEL="verbose"
```

Or, in PowerShell:

```bash
$env:AZURE_LOG_LEVEL="verbose"
```

> Then run your application as normal.  The logs will be emitted on the console.  For more information on the Java logging system, refer to [our wiki](https://github.com/Azure/azure-sdk-for-java/wiki/Logging-with-Azure-SDK).

Reviewer 4: ```suggestion
Then run your application as normal.  The logs will be emitted on the console.  Beginning with Azure Core 1.3.0-beta.1 a default implementation will be available to emit logs to the console. For more information on the Java logging system, refer to [our wiki](https://github.com/Azure/azure-sdk-for-java/wiki/Logging-with-Azure-SDK).
```

A default implementation is available in a preview release.

---
title: Azure SDK Releases (February 2020)
layout: post
date: 11 Feb 2020
sidebar: releases_sidebar
repository: azure/azure-sdk
---

The new generally available libraries being released this month are:

> * [Azure App Configuration](https://docs.microsoft.com/azure/azure-app-configuration/)

Reviewer 4: I thought this was shipped in January?

Author: Ignore - these will be updated after the shipping is closed out.

> * [Azure Key Vault Certificates](https://docs.microsoft.com/en-us/azure/key-vault/certificate-scenarios)

Reviewer 4: I thought this was shipped in January?

Author: Ditto - ignore - the top section just isn't updated yet.

## Logging 

One of the main advantages of the new Azure SDK is that common developer concerns are treated the same, irrespective of which client library you are using.  In this issue, we will tackle logging and show you how to enable logging on each platform.  

## .NET libraries

All .NET client libraries emit events to ETW (Event Tracing for Windows) via the `EventSource` class.  This system has been a part of the .NET framework for a long time.  Event sources allow you to use structured logging in your application code with a minimal performance overhead.

Although you can use out-of-process tools (such as **TODO: INSERT TOOLS HERE WITH LINKS**), a [core tenet of our libraries](https://azure.github.io/azure-sdk/general_implementation.html#general-logging-console-logger) includes the ability to easily send logs to the console.  When developing your app, you should be able to view the logs in real time without much overhead.  You can accomplish this with a one-liner at the top of your application:

```csharp
using AzureEventSourceListener listener = AzureEventSourceListener.CreateConsoleLogger();
```

## Java libraries

> The Java client libraries all use [SLF4J](https://www.slf4j.org/) under the covers for logging.  There are many mechanisms for [configuring SLF4J](https://dzone.com/articles/how-configure-slf4j-different) to get just the right logging for you application.  However, we are looking at the simplest mechanism todya.  There is a built-in logger for temporary debugging purposes.  As long as you have not already included the SLF4J library, just set the `AZURE_LOG_LEVEL` environment variable to "debug".  For example, in bash:

Reviewer 4: ```suggestion
The Java client libraries all use [SLF4J](https://www.slf4j.org/) under the covers for logging.  There are many mechanisms for [configuring SLF4J](https://dzone.com/articles/how-configure-slf4j-different) to get just the right logging for you application.  However, we are looking at the simplest mechanism today.  There is a built-in logger for temporary debugging purposes.  As long as you have not already included the SLF4J library, just set the `AZURE_LOG_LEVEL` environment variable to "debug".  For example, in bash:
```

> The Java client libraries all use [SLF4J](https://www.slf4j.org/) under the covers for logging.  There are many mechanisms for [configuring SLF4J](https://dzone.com/articles/how-configure-slf4j-different) to get just the right logging for you application.  However, we are looking at the simplest mechanism todya.  There is a built-in logger for temporary debugging purposes.  As long as you have not already included the SLF4J library, just set the `AZURE_LOG_LEVEL` environment variable to "debug".  For example, in bash:

Reviewer 4: The default logger hasn't been shipped yet and is unlikely to ship this month, do we want to annotate this message to state we are investigating the addition of a default logger likely to ship in the near term future.

Author: No - we want to rip it out and put in whatever it does look like.  What does the "simplest" case look like?

Author: Also - update the wiki - it's wrong if it hasn't shipped yet.

```

Or, in PowerShell:

```bash
$env:AZURE_LOG_LEVEL="debug"
```

Then run your application as normal.  The logs will be emitted on the console.  For more information on the Java logging system, refer to [our wiki](https://github.com/Azure/azure-sdk-for-java/wiki/Logging-with-Azure-SDK).

## JavaScript and TypeScript libraries

**TODO: ADD JAVASCRIPT CODE HERE**

## Python libraries

Python uses the [standard logging module](https://docs.python.org/3/howto/logging.html).  This makes it relly easy to configure just like you would any other Python library:

```python
import logging
logging.basicConfig()

# Enable DEBUG logging for all azure libraries
azure_root_logger = logging.getLogger('azure')
azure_root_logger.setLevel(logging.DEBUG)
```

You can see more configuration examples in the [logging cookbook](https://docs.python.org/3/howto/logging-cookbook.html).

As you can see, the same logging features are provided in each language, but how each language accomplishes them is idiomatic to the language.  How you work with these features should feel very natural in your language of choice.

If it doesn't, let us know!

## Working with us and giving feedback

> So far, the community has filed hundreds of issues against these new SDKs with feedback randing from documentation issues to API surface area change requests to pointing out failure cases.  Please keep that coming.  We work in the open on GitHub and you can submit issues here:

Reviewer 4: ```suggestion
So far, the community has filed hundreds of issues against these new SDKs with feedback ranging from documentation issues to API surface area change requests to pointing out failure cases.  Please keep that coming.  We work in the open on GitHub and you can submit issues here:
```
